### ISSUE TITLE
starting ranger with binding in bash/fish will cause expected behavior for pasting string when renaming files

### ISSUE THREAD

**Issue by c02y** [ranger/ranger#2287]
#### Runtime Environment

- Operating system and version: Manjaro-Linux
- Terminal emulator and version: xfce4-terminal-v0.8.10, alacritty-v0.7.2, kitty-v0.19.3
- ranger version: ranger-master
- Python version: 3.9.2 (default, Feb 20 2021, 18:40:11) [GCC 10.2.0]
- Locale: en_US.UTF-8

#### Current Behavior

When I bind a key in bash/fish such as `C-g` to start ranger, using default ranger config, `I` to rename a file, paste a string such as `abc` in you clipboard into the ranger command prompt, the pasted string will be like `[200~abc[201~`, but if I start ranger with typing in the shell prompt, it works fine.

#### Expected Behavior
When start ranger with the binding key, the pasted string should be the same as the original string in the clipboard.

#### Steps to reproduce
1. start bash, `bind -x '"\C-g": ranger'`, start ranger using `C-g` key
2. `I` to rename a file
3. `Shift-Insert` to insert a string in your clipboard
4. check the result

If you are using fish, use `bind \cg ranger`

#### Traceback

```

```

**Comment by toonn** [comment]
I think this is related to something the shell does differently. I don't know how to find out what. We need some shell experts on this.

**Comment by c02y** [comment]
fish shell is my daily shell, this issue existed for a long time, when I noticed this at the first beginning, I thought this is the problem of fish shell, then today I tried it in bash, it also happens in bash, so I guess this is the problem of ranger.

BTW: there is another program having the binding issue in fish(I didn't try it in bash, and forget what is the program and the scenario), so I really don't know what the root cause is.

**Comment by toonn** [comment]
Did you start bash from fish? Cause then it inherits some behavior.

**Comment by c02y** [comment]
No, I didn't start bash from fish, fish shell is my daily shell but it is not my default shell.

**Comment by c02y** [comment]
**UPDATE:**
Another problem has the binding problem https://github.com/jesseduffield/lazygit/issues/1179#issuecomment-812697885
But it is written in Go

**Comment by Enzime** [comment]
I believe this is a result of bracketed paste mode: https://en.wikipedia.org/wiki/ANSI_escape_code#cite_ref-26

Some resources I found:
- https://cirw.in/blog/bracketed-paste
- https://p.janouch.name/article-tui.html#Clipboard (this article talks more specifically about `ncurses`)

The simplest way is probably to just disable bracketed paste with `\e[?2004l`. If that doesn't work, possibly just eat the `\e[200~` and `\e[201~` when pasting, this would replicate the disabled bracketed paste mode. I'm not sure if `ranger` can handle newlines in the console at all.

**Comment by toonn** [comment]
@Enzime, any idea why starting ranger with a shell keybind would enable the mode?

**Comment by Enzime** [comment]
I've been encountering this same bug under `zsh` for a long time too and based on the text here: https://github.com/zsh-users/zsh/blob/68405f31a043bdd5bf338eb06688ed3e1f740937/README#L38-L45

I'm guessing that this is implemented specifically for `zle` (the `zsh` line editor), as this is what draws the prompt and takes normal input that people usually paste into and is where you want to protect from malicious pastes. This behaviour seems to be consistent with `bash` and `fish` as reported (I'm not too familiar those shells). However I'm not exactly sure why bracketed paste isn't being enabled when `ranger` is being run outside of the `zle`.

Here's an example of someone solving this from their config level: https://github.com/peco/peco/issues/417#issuecomment-289290816

**Comment by toonn** [comment]
@Enzime, are there other examples of TUI programs messing with bracketed paste mode? Should we always disable it on startup and try to reenable it on exit?

**Comment by c02y** [comment]
> 

Is it good to `__fish_disable_bracketed_paste` for the whole fish shell session? Since I sometimes `C-v` fails to paste the latest content from my clipboard (I use copyq), then I trigger copyq pop list and `Enter` on the right item in the list, there will be some `\e[200~` + the pasted item in the prompt at the same time.
